coderd/workspaceupdates.go

```diff
@@ -0,0 +1,313 @@
+	GetAuthorizedWorkspacesAndAgentsByOwnerID(ctx context.Context, ownerID uuid.UUID, prep rbac.PreparedAuthorized) ([]database.GetWorkspacesAndAgentsByOwnerIDRow, error)
+	GetWorkspaceByAgentID(ctx context.Context, agentID uuid.UUID) (database.Workspace, error)
+}
+
+type workspacesByID = map[uuid.UUID]ownedWorkspace
+
+type ownedWorkspace struct {
+	WorkspaceName string
+	Status        proto.Workspace_Status
+	Agents        []database.AgentIDNamePair
+}
+
+// Equal does not compare agents
+func (w ownedWorkspace) Equal(other ownedWorkspace) bool {
+	return w.WorkspaceName == other.WorkspaceName &&
+		w.Status == other.Status
+}
+
+type sub struct {
+	ctx      context.Context
+	cancelFn context.CancelFunc
+
+	mu       sync.RWMutex
+	userID   uuid.UUID
+	ch       chan *proto.WorkspaceUpdate
+	prev     workspacesByID
+	readPrep rbac.PreparedAuthorized
+
+	db     UpdatesQuerier
+	ps     pubsub.Pubsub
+	logger slog.Logger
+
+	psCancelFn func()
+}
+
+func (s *sub) handleEvent(ctx context.Context, event wspubsub.WorkspaceEvent, err error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	switch event.Kind {
+	case wspubsub.WorkspaceEventKindStateChange:
+	case wspubsub.WorkspaceEventKindAgentConnectionUpdate:
+	case wspubsub.WorkspaceEventKindAgentTimeout:
+	case wspubsub.WorkspaceEventKindAgentLifecycleUpdate:
+	default:
+		if err == nil {
+			return
+		} else {
+			// Always attempt an update if the pubsub lost connection
+			s.logger.Warn(ctx, "failed to handle workspace event", slog.Error(err))
+		}
+	}
+
+	rows, err := s.db.GetAuthorizedWorkspacesAndAgentsByOwnerID(ctx, s.userID, s.readPrep)
+	if err != nil {
+		s.logger.Warn(ctx, "failed to get workspaces and agents by owner ID", slog.Error(err))
+		return
+	}
+	latest := convertRows(rows)
+
+	out, updated := produceUpdate(s.prev, latest)
+	if !updated {
+		return
+	}
+
+	s.prev = latest
+	select {
+	case <-s.ctx.Done():
+		return
+	case s.ch <- out:
+	}
+}
+
+func (s *sub) start(ctx context.Context) (err error) {
+	rows, err := s.db.GetAuthorizedWorkspacesAndAgentsByOwnerID(ctx, s.userID, s.readPrep)
+	if err != nil {
+		return xerrors.Errorf("get workspaces and agents by owner ID: %w", err)
+	}
+
+	latest := convertRows(rows)
+	initUpdate, _ := produceUpdate(workspacesByID{}, latest)
+	s.ch <- initUpdate
+	s.prev = latest
+
+	cancel, err := s.ps.SubscribeWithErr(wspubsub.WorkspaceEventChannel(s.userID), wspubsub.HandleWorkspaceEvent(s.handleEvent))
+	if err != nil {
+		return xerrors.Errorf("subscribe to workspace event channel: %w", err)
+	}
+
+	s.psCancelFn = cancel
+	return nil
+}
+
+func (s *sub) Close() error {
+	s.cancelFn()
+
+	if s.psCancelFn != nil {
+		s.psCancelFn()
+	}
+
+	close(s.ch)
+	return nil
+}
+
+func (s *sub) Updates() <-chan *proto.WorkspaceUpdate {
+	return s.ch
+}
+
+var _ tailnet.Subscription = (*sub)(nil)
+
+type updatesProvider struct {
+	ps     pubsub.Pubsub
+	logger slog.Logger
+	db     UpdatesQuerier
+	auth   rbac.Authorizer
+
+	ctx      context.Context
+	cancelFn func()
```

Comment:
`ctx` isn't used anywhere, so it and `cancelFn` are superfluous, which I think makes `Close()` on this superfluous.